contributing/topics/defaults-and-breaking-changes.md

```diff
@@ -0,0 +1,128 @@
+
+We have a property like the following and Azure added a new spark version `3.4` and said that all new resources being created will be created with `3.4` as the default. 
+
+In Terraform, we start with:
+
+```hcl
+    "spark_version": {
+		Type:     pluginsdk.TypeString,
+		Optional: true,
+		Default:  "2.4",
+		ValidateFunc: validation.StringInSlice([]string{
+			"2.4",
+			"3.1",
+			"3.2",
+			"3.3",
+		}, false),
+	},
+```
+
+Then we would want to update `ValidateFunc` to include the new accepted value and update `Default` to `3.4` to keep it in line with Azure like so:
+
+```hcl
+    "spark_version": {
+		Type:     pluginsdk.TypeString,
+		Optional: true,
+		Default:  "3.4",
+		ValidateFunc: validation.StringInSlice([]string{
+			"2.4",
+			"3.1",
+			"3.2",
+			"3.3",
+			"3.4",
+		}, false),
+	},
+```
+
+But if we do that, people who have created that resource without the attribute specified will see a plan diff when upgrading to this version of the provider like so:
+
+This config does not specify `spark_version` because we know we can rely on the default to fill it for us
+
+```hcl
+resource "azurerm_synapse_spark_pool" "example" {
+  name                 = "example"
+  synapse_workspace_id = azurerm_synapse_workspace.test.id
+  node_size_family     = "MemoryOptimized"
+  node_size            = "Small"
+  node_count           = 3
+}
+```
+
+Running `terraform show` we can see `spark_version` has been filled in with the default of `2.4`
+
+```hcl
+# azurerm_synapse_spark_pool.example:
+resource "azurerm_synapse_spark_pool" "example" {
+    name                                = "example"
+.
+.
+.
+    spark_version                       = "2.4"
+}
+```
+
+When running the version of the provider where the default has changed from `2.4` to `3.4`, we'll see the following plan:
+
+```hcl
+Terraform will perform the following actions:
+
+  # azurerm_synapse_spark_pool.example will be updated in-place
+  ~ resource "azurerm_synapse_spark_pool" "test" {
+        id                                  = "exampleid"
+        name                                = "example"
+      ~ spark_version                       = "2.4" -> "3.4"
+        tags                                = {}
+        # (12 unchanged attributes hidden)
+    }
+
+Plan: 0 to add, 1 to change, 0 to destroy.
+```
+
+This is a breaking change as Terraform should not trigger a plan between minor version upgrades. Instead, what we can do is add a TODO next to the `Default` tag to update the default value in the next major version of the provider or mark the field as Required if that default value is going to continue to change in the future:
+
+```hcl
+    "spark_version": {
+		Type:     pluginsdk.TypeString,
+		Optional: true,
+		Default:  "2.4", // TODO: make this field Required in 4.0 since the default value changes over time
+		ValidateFunc: validation.StringInSlice([]string{
+			"2.4",
+			"3.1",
+			"3.2",
+			"3.3",
+			"3.4", 
+		}, false),
+	},
+```
```

Comment:
I think the preference here is to implement that change but using the `FourPointOhBeta` flag instead of a comment. I would only consider using a `// TODO 4.0` comment if the contributor is really unfamiliar with the code base and our practices.
```suggestion
This is a breaking change as Terraform should not trigger a plan between minor version upgrades. Instead, what we can do is add a TODO next to the `Default` tag to update the default value in the next major version of the provider or mark the field as Required if that default value is going to continue to change in the future:

```hcl
    "spark_version": {
		Type:     pluginsdk.TypeString,
		Optional: true,
		Default: func() string {
					if !features.FourPointOh() {
						return "2.4"
					}
					return "3.4"
				}(),
		ValidateFunc: validation.StringInSlice([]string{
			"2.4",
			"3.1",
			"3.2",
			"3.3",
			"3.4", 
		}, false),
	},
```

Reply:
Ooo I like that. Ty ty